File: holmesgpt-api/src/extensions/health.py
#
# Copyright 2025 Jordi Gil.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

"""
Health and Readiness Endpoints

Business Requirements: BR-HAPI-126 to 145 (Health/Monitoring)
"""

import importlib.util
import logging
from typing import Dict, Any
from fastapi import APIRouter, status
from fastapi.responses import JSONResponse

logger = logging.getLogger(__name__)

router = APIRouter()


def _check_sdk() -> bool:
    """
    Check if HolmesGPT SDK is available

    REFACTOR phase: Real SDK availability check
    Checks if the holmes package can be imported (installed via pip)
    """
    try:
        return importlib.util.find_spec("holmes") is not None
    except (ImportError, ModuleNotFoundError):
        logger.warning({"event": "sdk_not_found", "reason": "holmes package not importable"})
        return False
    except Exception as e:
        logger.error({"event": "sdk_check_failed", "error": str(e)})
        return False


def _check_prometheus() -> bool:
    """
    Check if Prometheus integration is available

    REFACTOR phase: Real Prometheus availability check
    """
    try:
        return importlib.util.find_spec("prometheus_client") is not None
    except (ImportError, ModuleNotFoundError):
        logger.warning({"event": "prometheus_not_available"})
        return False


def _check_context_api() -> bool:
    """
    Check if Context API is reachable

    REFACTOR phase: Real Context API connectivity check
    GREEN phase: Assume available (network policies handle access)
    """
    # For minimal service, assume available if running in same namespace
    # Network policies ensure connectivity
    return True


def _check_dependencies() -> Dict[str, Any]:
    """
    Check critical dependencies

    Business Requirement: BR-HAPI-017 (Dependency health checking)

    REFACTOR phase: Real dependency health checks
    """
    return {
        "sdk": {"status": "up" if _check_sdk() else "down"},
        "context_api": {"status": "up" if _check_context_api() else "down"},
        "prometheus": {"status": "up" if _check_prometheus() else "down"}
    }


@router.get("/health", status_code=status.HTTP_200_OK)
async def health_check():
    """
    Liveness probe endpoint

    Business Requirement: BR-HAPI-126 (Health check endpoint)
    """
    logger.info({"event": "health_check_called"})
    return {
        "status": "healthy",
        "service": "holmesgpt-api",
        "endpoints": [
            "/api/v1/incident/analyze",
            # DD-017: PostExec endpoint deferred to V1.1 - Effectiveness Monitor not in V1.0
            # "/api/v1/postexec/analyze",
            "/health",
            "/ready"
        ],
        "features": {
            "incident_analysis": True,
            # DD-017: PostExec feature deferred to V1.1 - Effectiveness Monitor not in V1.0
            # "postexec_analysis": True,
            "authentication": True
        },
    }


@router.get("/ready", status_code=status.HTTP_200_OK)
async def readiness_check():
    """
    Readiness probe endpoint

    Business Requirements:
    - BR-HAPI-127 (Readiness check endpoint)
    - BR-HAPI-201 (Graceful shutdown with DD-007 pattern)

    TDD GREEN Phase: Check shutdown flag first
    REFACTOR phase: Real dependency health checks
    """
    # Import shutdown flag from main module
    # BR-HAPI-201: Return 503 during graceful shutdown
    import src.main as main_module

    # Check if service is shutting down (DD-007 readiness coordination)
    if hasattr(main_module, 'is_shutting_down') and main_module.is_shutting_down:
        logger.info({
            "event": "readiness_check_shutting_down",
            "dd": "DD-007-readiness-503"
        })
        return JSONResponse(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            content={
                "status": "shutting_down",
                "reason": "Service is gracefully shutting down"
            }
        )

    # Check dependencies
    dependencies = _check_dependencies()

    # Check if all critical dependencies are up
    all_healthy = all(
        dep["status"] == "up"
        for dep in dependencies.values()
    )

    if not all_healthy:
        logger.warning({
            "event": "readiness_check_failed",
            "dependencies": dependencies
        })
        return JSONResponse(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            content={
                "status": "not_ready",
                "dependencies": dependencies
            }
        )

    logger.info({"event": "readiness_check_passed"})
    return {
        "status": "ready",
        "dependencies": dependencies
    }


@router.get("/config", status_code=status.HTTP_200_OK)
async def get_config():
    """
    Get service configuration (sanitized)

    Business Requirement: BR-HAPI-128 (Configuration endpoint)
    """
    config = getattr(router, "config", {})
    safe_config = {
        "llm": {
            "provider": config.get("llm", {}).get("provider"),
            "model": config.get("llm", {}).get("model"),
            "endpoint": config.get("llm", {}).get("endpoint")
        },
        "environment": config.get("environment"),
        "dev_mode": config.get("dev_mode"),
    }
    return safe_config

File: cmd/kubernautagent/health.go
/*
Copyright 2026 Jordi Gil.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/pprof"
	"os"
	"strings"
	"sync/atomic"
	"time"

	"github.com/go-logr/logr"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"go.uber.org/zap"
	"k8s.io/apimachinery/pkg/api/meta"
	"k8s.io/client-go/discovery/cached/memory"
	"k8s.io/client-go/dynamic"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/restmapper"
	ctrl "sigs.k8s.io/controller-runtime"

	kaconfig "github.com/jordigilh/kubernaut/internal/kubernautagent/config"
	karbac "github.com/jordigilh/kubernaut/internal/kubernautagent/rbac"
	"github.com/jordigilh/kubernaut/internal/kubernautagent/workflowcatalog"
	"github.com/jordigilh/kubernaut/internal/version"
	"github.com/jordigilh/kubernaut/pkg/fleet/readiness"
	"github.com/jordigilh/kubernaut/pkg/kubernautagent/llm"
)

func shutdownTimeout(cfg *kaconfig.Config) time.Duration {
	if cfg.Runtime.Shutdown.DrainSeconds > 0 {
		return time.Duration(cfg.Runtime.Shutdown.DrainSeconds) * time.Second
	}
	return 30 * time.Second
}

func shutdownServer(ctx context.Context, srv *http.Server, name string, logger logr.Logger) {
	if err := srv.Shutdown(ctx); err != nil {
		logger.Error(err, "server shutdown error", "server", name)
	}
}

func healthHandler(w http.ResponseWriter, _ *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(map[string]string{
		"status":    "healthy",
		"timestamp": time.Now().UTC().Format(time.RFC3339),
	})
}

// readinessHandler returns an http.HandlerFunc that performs real
// dependency checks instead of returning a static 200. Checks:
//   - shutdownFlag: set after SIGTERM/SIGINT received; makes probe fail
//     so k8s stops routing traffic during graceful shutdown.
//   - apiServerReady: set once the main API server goroutine is about to
//     start listening. Guards against the readiness probe reporting ready
//     before the main API server (auth middleware + JWKS pre-warm) has
//     finished initializing, which would otherwise let traffic hit a port
//     that isn't accepting connections yet.
//   - swappable: verifies the LLM client has a non-empty model name
//     (proxy for "LLM client was successfully initialized").
//   - ds: if non-nil, verifies the ogen client is initialized (DS is
//     a soft dependency — nil ds means DS is intentionally unconfigured).
//   - interactive: reports the interactive mode status (#891). This is
//     informational (does not fail the probe) since autonomous mode
//     continues to function even when interactive is soft-disabled.
//   - fleetGate: if non-nil, verifies the Fleet MCP Gateway is reachable
//     (#1553, ADR-068 decision #11). Nil means fleet mode is not
//     configured (soft dependency, matches the ds nil-check convention).
//   - wfCatalog: verifies KA's workflow catalog cache has completed its
//     first successful sync (#1677 hardening, DD-WORKFLOW-019). Unlike ds/
//     fleetGate, this is a hard (non-optional) dependency -- KA always runs
//     in-cluster and always constructs a wfCatalog, so nil or Not-Ready
//     here always fails the probe, keeping the pod out of Service
//     endpoints until discovery is genuinely available.
//   - dsGate: verifies DataStorage itself is reachable (#1985,
//     BR-AUDIT-005 v2.0). Unlike fleetGate, this is a hard (non-optional,
//     always non-nil) dependency -- every service writes audit, so a
//     Not-Ready DataStorage always fails the probe, closing the
//     audit-loss window where a pod accepts traffic before DataStorage is
//     confirmed reachable.
//
// readinessDeps groups readinessHandler's dependencies (Go anti-pattern
// checklist: 8+ parameters -> config struct instead of a long positional
// argument list).
type readinessDeps struct {
	shutdownFlag   *int32
	apiServerReady *int32
	swappable      *llm.SwappableClient
	ds             *dsClients
	interactive    *karbac.InteractiveReadiness
	fleetGate      *readiness.Gate
	wfCatalog      *workflowcatalog.LazyCatalog
	dsGate         *readiness.Gate
}

func readinessHandler(deps readinessDeps) http.HandlerFunc {
	return func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("Content-Type", "application/json")

		if atomic.LoadInt32(deps.shutdownFlag) != 0 {
			w.WriteHeader(http.StatusServiceUnavailable)
			_ = json.NewEncoder(w).Encode(map[string]string{
				"status": "not_ready",
				"reason": "shutting_down",
			})
			return
		}

		if atomic.LoadInt32(deps.apiServerReady) == 0 {
			w.WriteHeader(http.StatusServiceUnavailable)
			_ = json.NewEncoder(w).Encode(map[string]string{
				"status": "not_ready",
				"reason": "api_server_starting",
			})
			return
		}

		if deps.swappable.ModelName() == "" {
			w.WriteHeader(http.StatusServiceUnavailable)
			_ = json.NewEncoder(w).Encode(map[string]string{
				"status": "not_ready",
				"reason": "llm_client_not_configured",
			})
			return
		}

		if deps.ds != nil && deps.ds.ogenClient == nil {
			w.WriteHeader(http.StatusServiceUnavailable)
			_ = json.NewEncoder(w).Encode(map[string]string{
				"status": "not_ready",
				"reason": "datastorage_client_unavailable",
			})
			return
		}

		if deps.fleetGate != nil && !deps.fleetGate.Ready() {
			w.WriteHeader(http.StatusServiceUnavailable)
			_ = json.NewEncoder(w).Encode(map[string]string{
				"status": "not_ready",
				"reason": "fleet_mcp_gateway_unreachable",
			})
			return
		}

		if deps.wfCatalog == nil || !deps.wfCatalog.Ready() {
			w.WriteHeader(http.StatusServiceUnavailable)
			_ = json.NewEncoder(w).Encode(map[string]string{
				"status": "not_ready",
				"reason": "workflow_catalog_not_ready",
			})
			return
		}

		if deps.dsGate != nil && !deps.dsGate.Ready() {
			w.WriteHeader(http.StatusServiceUnavailable)
			_ = json.NewEncoder(w).Encode(map[string]string{
				"status": "not_ready",
				"reason": "datastorage_unreachable",
			})
			return
		}

		resp := map[string]string{
			"status":           "ready",
			"interactive_mode": deps.interactive.StatusString(),
		}
		if reason := deps.interactive.Reason(); reason != "" {
			resp["interactive_reason"] = reason
		}
		_ = json.NewEncoder(w).Encode(resp)
	}
}

func configHandler(cfg *kaconfig.Config, swappable *llm.SwappableClient) http.HandlerFunc {
	return func(w http.ResponseWriter, _ *http.Request) {
		model := swappable.ModelName()
		sanitized := map[string]interface{}{
			"service":     "kubernaut-agent",
			"version":     version.Version,
			"llm_model":   model,
			"session_ttl": cfg.Runtime.Session.TTL.String(),
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(sanitized)
	}
}

// healthServersParams groups the dependencies needed to build and start the
// health/readiness/metrics HTTP servers. Extracted per AGENTS.md's 8+-param
// Options-pattern rule.
type healthServersParams struct {
	Config               *kaconfig.Config
	AtomicLevel          zap.AtomicLevel
	Swappable            *llm.SwappableClient
	DS                   *dsClients
	InteractiveReadiness *karbac.InteractiveReadiness
	ShutdownFlag         *int32
	APIServerReady       *int32
	FleetGate            *readiness.Gate
	WfCatalog            *workflowcatalog.LazyCatalog
	// DSGate is the #1985 DataStorage readiness gate (BR-AUDIT-005 v2.0),
	// always non-nil in production (unlike FleetGate).
	DSGate *readiness.Gate
	Logger logr.Logger
}

// startHealthAndMetricsServers builds and starts (in background goroutines)
// the health/readiness/config/admin HTTP server and the Prometheus metrics
// server. These are started before API route setup so liveness/readiness
// probes are served even while the JWKS pre-warm (up to 15s) blocks inside
// newAuthMiddleware — otherwise the liveness probe kills the pod before the
// health server ever starts.
func startHealthAndMetricsServers(p healthServersParams) (*http.Server, *http.Server) {
	cfg, atomicLevel, swappable, ds, interactiveReadiness, shutdownFlag, apiServerReady, logger :=
		p.Config, p.AtomicLevel, p.Swappable, p.DS, p.InteractiveReadiness, p.ShutdownFlag, p.APIServerReady, p.Logger

	healthMux := http.NewServeMux()
	healthMux.HandleFunc("/healthz", healthHandler)
	healthMux.HandleFunc("/readyz", readinessHandler(readinessDeps{
		shutdownFlag:   shutdownFlag,
		apiServerReady: apiServerReady,
		swappable:      swappable,
		ds:             ds,
		interactive:    interactiveReadiness,
		fleetGate:      p.FleetGate,
		wfCatalog:      p.WfCatalog,
		dsGate:         p.DSGate,
	}))
	healthMux.HandleFunc("/config", configHandler(cfg, swappable))
	if !cfg.Runtime.Server.DisableAdminEndpoints {
		healthMux.Handle("/admin/loglevel", atomicLevel)
	}
	if cfg.Runtime.Debug.PprofEnabled {
		healthMux.HandleFunc("/debug/pprof/", pprof.Index)
		healthMux.HandleFunc("/debug/pprof/cmdline", pprof.Cmdline)
		healthMux.HandleFunc("/debug/pprof/profile", pprof.Profile)
		healthMux.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
		healthMux.HandleFunc("/debug/pprof/trace", pprof.Trace)
	}
	healthServer := &http.Server{
		Addr:              cfg.Runtime.Server.HealthAddr,
		Handler:           healthMux,
		ReadHeaderTimeout: 5 * time.Second,
		ReadTimeout:       10 * time.Second,
		WriteTimeout:      10 * time.Second,
	}
	metricsMux := http.NewServeMux()
	metricsMux.Handle("/metrics", promhttp.Handler())
	metricsServer := &http.Server{
		Addr:              cfg.Runtime.Server.MetricsAddr,
		Handler:           metricsMux,
		ReadHeaderTimeout: 5 * time.Second,
		ReadTimeout:       5 * time.Second,
		WriteTimeout:      10 * time.Second,
		IdleTimeout:       30 * time.Second,
	}

	go func() {
		logger.Info("health server listening", "addr", cfg.Runtime.Server.HealthAddr)
		if err := healthServer.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			logger.Error(err, "health server error")
		}
	}()
	go func() {
		logger.Info("metrics server listening", "addr", cfg.Runtime.Server.MetricsAddr)
		if err := metricsServer.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			logger.Error(err, "metrics server error")
		}
	}()

	return healthServer, metricsServer
}

// detectNamespace reads the pod's namespace from the mounted ServiceAccount.
//
// KUBERNAUT_AGENT_NAMESPACE takes precedence when set: this lets multiple
// per-process KA instances sharing a single envtest apiserver (AA IT
// shared-envtest fix, DD-TEST-010 amendment) each be scoped to their own
// process-unique namespace instead of racing over the same hardcoded
// default. Empty/unset preserves production behavior exactly.
func detectNamespace() string {
	if ns := os.Getenv("KUBERNAUT_AGENT_NAMESPACE"); ns != "" {
		return ns
	}
	data, err := os.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
	if err == nil && len(data) > 0 {
		return strings.TrimSpace(string(data))
	}
	return "kubernaut-system"
}

// k8sInfra holds shared Kubernetes clients created once and reused by
// the tool registry, enricher, and custom tools.
type k8sInfra struct {
	kubeConfig *rest.Config
	clientset  *kubernetes.Clientset
	dynClient  dynamic.Interface
	mapper     meta.RESTMapper
}

// CRITICAL: match DataStorage's (cmd/datastorage/main.go) and Gateway's
// (test/integration/gateway/suite_test.go) already-established fix for the
// same failure mode: client-go's rest.Config defaults to QPS=5, Burst=10
// when unset, which silently queues (Wait()s, no error, no log line) once a
// client issues more concurrent requests than that budget allows.
//
// RCA (grounded in job logs + must-gather, PR #1739 CI run 30213808860, job
// 89825941993, issue #1741): E2E-AF-1637-001 intermittently failed because
// KA's investigator produced zero events for 15s after session start —
// mock-llm's own log confirms KA's first LLM call for that session didn't
// fire until exactly 15s after "launchInvestigation" — the classic silent
// symptom of this shared clientset/dynamic-client/RESTMapper (built here,
// consumed by enrichment Get() calls, owner-reference lookups, and
// RESTMapper-backed scope resolution) queuing behind concurrent
// investigations' K8s calls under this exact default. KA's OWN
// buildMCPControllerClient (routes.go) already overrides QPS/Burst for its
// controller-runtime client (SEC-07, deliberately conservative at 20/40 for
// that MCP-driven path) — this clientset/dynClient/mapper trio was the one
// client in this binary still left on the client-go default.
func k8sQPSBurst(cfg *rest.Config) *rest.Config {
	cfg.QPS = 1000.0
	cfg.Burst = 2000
	return cfg
}

// initK8sInfra creates the shared Kubernetes clients. Returns nil when
// running outside a cluster (e.g. local development).
func initK8sInfra(logger logr.Logger) *k8sInfra {
	kubeConfig, err := ctrl.GetConfig()
	if err != nil {
		logger.Info("K8s config not available, K8s tools and enricher disabled", "error", err)
		return nil
	}
	kubeConfig = k8sQPSBurst(kubeConfig)

	k8sClient, err := kubernetes.NewForConfig(kubeConfig)
	if err != nil {
		logger.Error(err, "failed to create K8s clientset")
		return nil
	}
	dynClient, err := dynamic.NewForConfig(kubeConfig)
	if err != nil {
		logger.Error(err, "failed to create dynamic client")
		return nil
	}
	cachedDisc := memory.NewMemCacheClient(k8sClient.Discovery())
	mapper := restmapper.NewDeferredDiscoveryRESTMapper(cachedDisc)
	return &k8sInfra{kubeConfig: kubeConfig, clientset: k8sClient, dynClient: dynClient, mapper: mapper}
}
